From 7d2e41d3bcbf23765efc01e55d73dda6c0bced78 Mon Sep 17 00:00:00 2001
From: Eric Garver <eric@garver.life>
Date: Tue, 3 Jun 2025 16:20:11 -0400
Subject: [PATCH 34/34] v2.4.0: test: rich: rule reference invalid ipset

Coverage: RHEL-67103
Coverage: RHEL-67331
(cherry picked from commit d99076a59692af306e4dd6c909351a69e80b082a)
---
 src/tests/regression/RHEL-67103.at | 34 ++++++++++++++++++++++++++++++
 src/tests/regression/regression.at |  1 +
 2 files changed, 35 insertions(+)
 create mode 100644 src/tests/regression/RHEL-67103.at

diff --git a/src/tests/regression/RHEL-67103.at b/src/tests/regression/RHEL-67103.at
new file mode 100644
index 000000000000..0cca717246bf
--- /dev/null
+++ b/src/tests/regression/RHEL-67103.at
@@ -0,0 +1,34 @@
+FWD_START_TEST([rich rule invalid ipset])
+AT_KEYWORDS(rich ipset RHEL-67103 RHEL-67331)
+
+dnl valid ipset
+FWD_CHECK([--permanent --new-ipset thisexists --type=hash:net], 0, [ignore])
+FWD_CHECK([--permanent --add-rich-rule="rule family=ipv4 source ipset=thisexists accept"], 0, [ignore])
+FWD_CHECK([--permanent --add-rich-rule="rule family=ipv4 destination ipset=thisexists accept"], 0, [ignore])
+FWD_RELOAD()
+
+dnl invalid ipset
+FWD_CHECK([            --add-rich-rule="rule family=ipv4 source ipset=doesnotexist accept"], 135, [ignore], [ignore])
+FWD_CHECK([--permanent --add-rich-rule="rule family=ipv4 source ipset=doesnotexist accept"], 135, [ignore], [ignore])
+FWD_CHECK([            --add-rich-rule="rule family=ipv4 destination ipset=doesnotexist accept"], 135, [ignore], [ignore])
+FWD_CHECK([--permanent --add-rich-rule="rule family=ipv4 destination ipset=doesnotexist accept"], 135, [ignore], [ignore])
+
+dnl verify checks for manually added XML
+AT_DATA([./zones/broken.xml], [m4_strip([dnl
+  <?xml version="1.0" encoding="utf-8"?>
+  <zone>
+    <short>broken</short>
+    <rule family="ipv4">
+      <source ipset="doesnotexist"/>
+      <accept/>
+    </rule>
+    <rule family="ipv4">
+      <destination ipset="doesnotexist"/>
+      <accept/>
+    </rule>
+    <forward/>
+  </zone>
+])])
+FWD_RELOAD(135, [ignore], [ignore])
+
+FWD_END_TEST([-e '/ERROR: INVALID_IPSET/d'])
diff --git a/src/tests/regression/regression.at b/src/tests/regression/regression.at
index e85641932b58..ea1357a91d44 100644
--- a/src/tests/regression/regression.at
+++ b/src/tests/regression/regression.at
@@ -57,3 +57,4 @@ m4_include([regression/ipset_scale.at])
 m4_include([regression/gh1129.at])
 m4_include([regression/gh1146.at])
 m4_include([regression/rhbz2222044.at])
+m4_include([regression/RHEL-67103.at])
-- 
2.47.1

